fendermint/vm/interpreter/src/fvm/checkpoint.rs

```diff
@@ -74,9 +76,19 @@ where
                 .context("failed to apply validator changes")?;
 
             // Retrieve the bottom-up messages so we can put their hash into the checkpoint.
-            let cross_messages_hash = gateway
-                .bottom_up_msgs_hash(state, height.value())
-                .context("failed to retrieve bottom-up messages hash")?;
+            let cross_msgs = gateway
+                .bottom_up_msgs(state, height.value())
+                .context("failed to retrieve bottom-up messages")?;
+
+            // Sum up the value leaving the subnet as part of the bottom-up messages.
+            let burnt_tokens = tokens_to_burn(&cross_msgs);
+
+            // TODO: Should we actually take this from the gateway balance? Or did it already get burnt?
```

Comment:
It should already be burnt and discussed here: https://github.com/consensus-shipyard/ipc-solidity-actors/pull/263

Reply:
Ah great, thanks for the link. I see you are sending to the `BURNT_FUNDS_ACTOR`. In that case we have two options:
* We don't adjust the circulating supply here, rather, we look up the balance of the burnt funds actor on each block and subtract it from the circulating supply stored in the `AppState`. That way _anybody_ can send funds to the burnt funds actor, not just IPC, and we'll get the correct value *(see caveat below). 
* We decrease the circulaing supply here which makes the one in `AppState` be in sync with IPC itself, however, the actual circulating supply can be lower because anyone else can burn funds by sending to the burnt funds actor, and now unfortunately we can't subtract it as it would go negative. 

*Caveat: both cases can misrepresent the actual circulating supply if people use a random account for burning funds rather than the well known burnt funds address.

Reply:
I think both options are good. If it is not hard to implement I personally like option 1. If we want to be super accurate, we could also check the balance for the EVM zero address and not only the `BURNT_FUNDS_ACTOR`. This gives us an IPC-agnostic solution :) 

That being said, I think this is the most accurate we could go, so I wouldn't worry about trying to be 100% accurate (not even Bitcoin has this knowledge, as there maybe funds lost forever on accounts for which private keys have been losted, and strictly speaking these don't belong to the circulating supply).

Reply:
I left it as it is, which means it's 100% accurate with what IPC is doing, making at least that one a known quantity.

Reply:
That's fine, I just realized one could always explicitly check the balances of those addresses if they want an accurate value.